### PR TITLE
Update the documentation of SmartPointer.

### PR DESCRIPTION
**Pull request by bangerth**
I had a conversation with one of my students about the class and thought it's worthwhile updating its documentation. I don't think we ever looked at it in the last 20 years.

### CODE REVIEW

File: include/deal.II/base/smartpointer.h [hunk 1]
```diff
@@ -27,39 +27,65 @@
 DEAL_II_NAMESPACE_OPEN
 
 /**
- * Smart pointers avoid using dangling pointers. They can be used just
+ * The SmartPointer class avoids using dangling pointers. They can be used just
  * like a pointer (i.e. using the <tt>*</tt> and <tt>-></tt> operators and
```
Comment (marcfehling) [inline include/deal.II/base/smartpointer.h:31]:
```suggestion
 * like a pointer (i.e., using the <tt>*</tt> and <tt>-></tt> operators and
```

File: include/deal.II/base/smartpointer.h [hunk 2]
```diff
@@ -27,39 +27,65 @@
...
- * T* p = new T;
- * {
- *   SmartPointer<T,P> t(p);
- *   ...
- * }
- * delete p;
- * @endcode
+ * To make this scheme work, SmartPointers need to increment the "observer
+ * count" when they point to an observed object. This is facilitated by
+ * requiring that the objects pointed to, i.e. the template type `T`, must
```
Comment (marcfehling) [inline include/deal.II/base/smartpointer.h:55]:
```suggestion
 * requiring that the objects pointed to, i.e., the template type `T`, must
```

File: include/deal.II/base/smartpointer.h [hunk 3]
```diff
@@ -27,39 +27,65 @@
...
+ *   take over ownership. As a consequence, this is a sure way of creating
+ *   a memory leak:
+ *   @code
+ *     SmartPointer<T> dont_do_this = new T;
+ *   @endcode
+ *   This is because here, no variable "owns" the object pointed to, and
+ *   the destruction of the `dont_do_this` pointer does not trigger the
+ *   release of the memory pointed to.
+ *
+ * @note This class correctly handles `const`-ness of an object, i.e.
```
Comment (marcfehling) [inline include/deal.II/base/smartpointer.h:85]:
```suggestion
 * @note This class correctly handles `const`-ness of an object, i.e.,
```

**Review by marcfehling — APPROVED** [review]
I like the changes!

Only suggestion I have is to use commas after `i.e.`

**Comment by bangerth** [comment]
So fixed.

**Review by marcfehling — APPROVED** [review]

**Review by kronbichler — APPROVED** [review]

**Pull request merged**